packages/common/src/lib/common/tools.ts

```diff
@@ -2184,44 +2243,40 @@ function promisifyNoError(fn, context, returnArgNames) {
 
 /**
  * Creates and executes an array of promises in sequence
- * @param {((...args: any[]) => Promise<any>)[]} promiseFactories An array of promise-returning functions
+ * @param promiseFactories An array of promise-returning functions
  */
-function promiseSequence(promiseFactories) {
+function promiseSequence(promiseFactories: ((...args: any[]) => Promise<any>)[]): any[] {
+    /** @ts-expect-error don't want to touch now */
     return promiseFactories.reduce((promise, factory) => {
         return promise.then(result => factory().then(Array.prototype.concat.bind(result)));
     }, Promise.resolve([]));
 }
 
-function _setQualityForStates(states, keys, quality, cb) {
-    if (!keys || !states || !keys.length) {
-        cb();
-    } else {
-        states.setState(
-            keys.shift(),
-            {
-                ack: null,
-                q: quality
-            },
-            () => setImmediate(_setQualityForStates, states, keys, quality, cb)
-        );
+async function _setQualityForStates(states: any, keys: string[], quality: number): Promise<void> {
+    for (const key of keys) {
+        try {
+            await states.setState(key, { ack: null, q: quality });
+        } catch {
+            // ignore
+        }
     }
 }
 
-function setQualityForInstance(objects, states, namespace, q) {
-    return new Promise((resolve, reject) => {
+function setQualityForInstance(objects: any, states: any, namespace: string, q: number): Promise<void> {
+    return new Promise<void>((resolve, reject) => {
         objects.getObjectView(
```

Comment:
Can be made async I think.

Reply:
yes it can but not necessarily in this PR :smile: